core/src/main/java/com/linecorp/armeria/client/limit/AsyncConcurrencyLimit.java

```diff
@@ -0,0 +1,247 @@
+
+import io.netty.channel.EventLoop;
+import io.netty.util.concurrent.Future;
+import io.netty.util.concurrent.FutureListener;
+import io.netty.util.concurrent.Promise;
+
+/**
+ * Concurrency settings that limits the concurrent number of active requests.
+ */
+final class AsyncConcurrencyLimit implements ConcurrencyLimit {
+    private final int maxParallelism;
+    private final int maxPendingAcquires;
+    private final long acquireTimeoutNanos;
+
+    private final Queue<AcquireTask> pendingAcquireQueue;
+    private final AtomicInteger acquiredPermitCount = new AtomicInteger();
+    private final Runnable timeoutTask;
+    private int pendingAcquireCount;
+    private final SafeCloseable semaphorePermit = this::decrementAndRunTaskQueue;
+
+    AsyncConcurrencyLimit(
+            long acquireTimeoutMillis,
+            int maxParallelism,
+            int maxPendingAcquires) {
+        this.maxParallelism = maxParallelism;
+        this.maxPendingAcquires = maxPendingAcquires;
+        this.pendingAcquireQueue = new ArrayDeque<>(maxPendingAcquires);
+        this.acquireTimeoutNanos = TimeUnit.MILLISECONDS.toNanos(acquireTimeoutMillis);
+        this.timeoutTask = new TimeoutTask();
+    }
+
+    /**
+     * Returns the total number of acquired permits.
+     */
+    public int acquiredPermitCount() {
+        return acquiredPermitCount.get();
+    }
+
+    /**
+     * Returns the total number of available permits.
+     */
+    public int availablePermitCount() {
+        return maxParallelism - acquiredPermitCount();
+    }
+
+    @Override
+    public CompletableFuture<SafeCloseable> acquire(ClientRequestContext requestContext) {
+        return Futures.toCompletableFuture(acquire(requestContext, requestContext.eventLoop().newPromise()));
+    }
+
+    private Future<SafeCloseable> acquire(final ClientRequestContext requestContext,
+                                          final Promise<SafeCloseable> promise) {
+        final EventLoop executor = requestContext.eventLoop().withoutContext();
+        try {
+            if (executor.inEventLoop()) {
+                acquire0(executor, promise);
+            } else {
+                executor.execute(() -> acquire0(executor, promise));
+            }
+        } catch (Throwable cause) {
+            promise.setFailure(cause);
+        }
+        return promise;
+    }
+
+    private void acquire0(EventLoop executor, final Promise<SafeCloseable> promise) {
+        assert executor.inEventLoop();
+
+        if (acquiredPermitCount.get() < maxParallelism) {
+            assert acquiredPermitCount.get() >= 0;
+
+            // We need to create a new promise as we need to ensure the AcquireListener runs in the correct
+            // EventLoop
+            final Promise<SafeCloseable> p = executor.newPromise();
+            final AcquireListener l = new AcquireListener(executor, promise);
+            l.acquired();
+            p.addListener(l);
+            newPermit(p);
+        } else {
```

Comment:
`AcquireListener` and the extra `Promise` don't seem to be needed at all here if I understood correctly. Why don't we just do what `AcquireListener.acquired()` and `.operationComplete()` here and then call `promise.setSuccess(semaphorePermit)`? i.e.

```java
            // Acquire immediately.
            acquiredPermitCount.incrementAndGet();
            promise.setSuccess(semaphorePermit);
```

Once simplified, we can merge `AcquireListener` into `AcquireTask` for further simplification.